src/Hangfire.Mongo/Signal/Mongo/MongoSignal.cs

```diff
@@ -0,0 +1,159 @@
+                        Builders<SignalDto>.Update.Set(s => s.Signaled, false),
+                        null,
+                        cancellationToken);
+                    if (signalDto == null)
+                    {
+                        Logger.Debug($@"*** TRIGGERED ({Thread.CurrentThread.ManagedThreadId})");
+                    }
+                    else
+                    {
+                        Logger.Debug($@"*** OWNED: {signalDto.Name} ({Thread.CurrentThread.ManagedThreadId})");
+                    }
+                    EventWaitHandles
+                        .GetOrAdd(document.Name, n => new EventWaitHandle(false, EventResetMode.AutoReset))
+                        .Set();
+                }, cancellationToken).ContinueWith(t =>
+                {
+                    // TODO: Log if canceled?
+                    if (t.IsFaulted)
+                    {
+                        var messages = string.Join(Environment.NewLine, t.Exception.InnerExceptions.Select(e => e.Message));
+                        Logger.Debug($@"*** FAULT: {messages} ({Thread.CurrentThread.ManagedThreadId})");
+                    }
+                }).Wait();
+            }
+        }
+
+        public void Set(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name should not be empty", nameof(name));
+            }
+
+            var update = Builders<SignalDto>.Update
+                .Set(s => s.Signaled, true)
+                .Set(s => s.Name, name)
+                .CurrentDate(s => s.TimeStamp);
+            _signal.UpdateOne(_ => false, update, new UpdateOptions { IsUpsert = true });
+
+            Logger.Debug($@"*** SET: {name} ({Thread.CurrentThread.ManagedThreadId})");
+        }
+
+        public void Wait(string name)
+        {
+            Wait(name, TimeSpan.MaxValue);
+        }
+
+        public void Wait(string name, TimeSpan timeout)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name should not be empty", nameof(name));
+            }
+
+            var e = EventWaitHandles.GetOrAdd(name, n => new EventWaitHandle(false, EventResetMode.AutoReset));
+            Logger.Debug($@">>> WAIT: {name} ({Thread.CurrentThread.ManagedThreadId})");
+            if (timeout == TimeSpan.MaxValue)
+            {
+                e.WaitOne();
+            }
+            else
+            {
+                e.WaitOne(timeout);
+            }
+            Logger.Debug($@"<<< WAIT: {name} ({Thread.CurrentThread.ManagedThreadId})");
+        }
+
+        public void Wait(string name, CancellationToken cancellationToken)
```

Comment:
Maybe use the cancellationTokens waithandle instead of using Task.Run?

Reply:
I'll look into that...

Reply:
No more Task.Run. Tnx for pointing out...